packages/vite/src/node/plugins/resolve.ts

```diff
@@ -783,11 +788,30 @@ function resolveExports(
   if (options.conditions) {
     conditions.push(...options.conditions)
   }
-  return _resolveExports(pkg, key, {
-    browser: targetWeb,
-    require: options.isRequire,
-    conditions
-  })
+
+  try {
+    const resolved = _resolveExports(pkg, keyWithQuery, {
+      browser: targetWeb,
+      require: options.isRequire,
+      conditions
+    })
+    return resolved
+  } catch (err) {
+    // not found
+
+    // try without postfix for `import 'something/path.js?query'` (see #7098)
+    const { file, postfix } = splitFileAndPostfix(keyWithQuery)
+    if (!postfix) {
+      throw err
+    }
+
+    const resolvedWithoutPostfix = _resolveExports(pkg, file, {
+      browser: targetWeb,
+      require: options.isRequire,
+      conditions
+    })
+    return resolvedWithoutPostfix + postfix
+  }
```

Comment:
I've been thinking about this for awhile, and I'm not sure why I suggested going with handling postfix for `exports` and `browser`. Sorry if I've mislead that way, but it doesn't seem right to me that we're handling these cases to for `exports` and `browser`, since the keys usually don't contain `#` and `?`, and we don't want to work for them too.

I'm hoping you can help me understand [your comment](https://github.com/vitejs/vite/pull/7098#discussion_r815335315), and whether we should proceed with a logic like this instead:

In `resolveDeepImport()`

1. If have `exports`
  i. strip `file` and `postfix` for `relativeId`
  ii. use `file` to match `resolveExports`
  iii. if there's an id returned, add back the `postfix` and return
2. Else if have `browser`
  i. strip `file` and `postfix` for `relativeId`
  ii. use `file` to match `mapWithBrowserField`
  iii. if there's an id returned, add back the `postfix` and return
3. Else for `fs`, leave the code as is

Let me know what do you think!

Reply:
When I created this PR, I thought it is not possible or not recommended to include `?` in exports field.
But after I got your review, I check the [docs](https://nodejs.org/api/packages.html#exports) and thought it is possible since it requires to be a relative file url. So I thought I should implement like I have done currently.  I was not sure it is possible with browser field but I expected to be same with exports.
(Never mind with the misleadings!)

To sum up, I think `?` and `#` should not be used with exports and browser field. Also the logic you suggested in this comment would be good.

Reply:
Awesome thanks for making the changes again. I saw the docs too and tested locally that `?` actually works in conditional exports, but I feel like that shouldn't be something Vite should be encouraging. Nonetheless, I'm taking another look with the changes now.